### ISSUE TITLE
Contrib project

### ISSUE THREAD

**Issue by cwe1ss** [opentracing/opentracing-csharp#29]
I finally started with a contrib-project! The repo is under my account [cwe1ss/opentracing-contrib-dotnet](https://github.com/cwe1ss/opentracing-contrib-dotnet) but it would be great if we could move this to [opentracing-contrib](https://github.com/opentracing-contrib) at some point.

Right now, it traces incoming HTTP requests in ASP.NET Core and outgoing HTTP requests with HttpClient.

It's not tested because there's no tracer implementation yet :smile:. Hopefully, I'll find some time over the next few weeks to work on BasicTracer and get a working integration with e.g. Zipkin. I can't wait to finally see a first trace in a GUI :smile:.

**Comment by dawallin** [comment]
@cwe1ss Nice work.
What do you think of creating a separate brand to host all these extra parts, together with all extension methods we removed from OT to keep API surface small. The list of surrounding projects will grow and I think it is not necessarily a bad thing to create a separate brand than opentracing around them. I think that it could both be an OSS implementation of the OpenTracing API, together with this type of adaptors. We would then be more flexible then using the open tracing in everything.
I came up with the name [MicroCrumbs](https://github.com/Microcrumbs) a while ago where I planed to put this kind of code. If you like it we can use it or if you have a better name I'm open for that as well.

**Comment by cwe1ss** [comment]
thanks - still a lot to do though! :smile:

There's the separate github organization [opentracing-contrib](https://github.com/opentracing-contrib) which is meant for language-specific projects. Being listed there would probably give us the best "marketing push" because the sad truth is, no-one knows about this in the .NET world yet. :(

My concerns with a separate name are that it will be harder for people to understand the relationship. By using `OpenTracing.Contrib`, we would share the name and we should pop up whenever someone is searching for opentracing and .NET. 
Also, the implementations are not really projects themselves - they just add instrumentation for existing libraries/frameworks. Having `OpenTracing.Contrib.AspNetCore` (or just `OpenTracing.AspNetCore` if we don't need the Contrib tag) would be easier to understand than `MicroCrumbs.AspNetCore`.

MicroCrumbs sounds great for a tracing system though! If you plan to build a .NET based tracer, I'd definitely vote for it :)

Your thoughts?

**Comment by cwe1ss** [comment]
/cc @adriancole @bensigelman @dawallin @yurishkuro 

I now have a first working "alpha" version of the following .NET / C# packages. Before we take a closer look at them and move them to opentracing-contrib, I'd like to get your feedback regarding the package names and some general things!

* [OpenTracing.Instrumentation](https://github.com/cwe1ss/opentracing-contrib-dotnet/tree/master/src/OpenTracing.Instrumentation): Targets library & application developers. It contains in-process trace propagation, helper methods and instrumentation for HTTP calls and database calls (done by an ORM tool). Thanks to some nice hooks in .NET, this works without package dependencies, so I included it in this package.
* [OpenTracing.Instrumentation.AspNetCore](https://github.com/cwe1ss/opentracing-contrib-dotnet/tree/master/src/OpenTracing.Instrumentation.AspNetCore): Instrumentation for the "ASP.NET Core" web framework (incoming requests, some MVC performance). Depends on `OpenTracing.Instrumentation`.
* [OpenTracing.Testing](https://github.com/cwe1ss/opentracing-contrib-dotnet/tree/master/src/OpenTracing.Testing): In-memory tracer for unit tests.
* [OpenTracing.Tracer.Abstractions](https://github.com/cwe1ss/opentracing-contrib-dotnet/tree/master/src/OpenTracing.Tracer.Abstractions): Contains base classes for tracers. Similar to the Java AbstractTracer. Ideally, this would also be used by the `BasicTracer`.
* [OpenTracing.Tracer.BatchReporter](https://github.com/cwe1ss/opentracing-contrib-dotnet/tree/master/src/OpenTracing.Tracer.BatchReporter): Reusable library for tracers that queues spans and allows tracers to send spans in batches on a background thread. 
* [OpenTracing.Tracer.Zipkin](https://github.com/cwe1ss/opentracing-contrib-dotnet/tree/master/src/OpenTracing.Tracer.Zipkin): A tracer that sends data to Zipkin. This currently is a mix of some existing Zipkin libraries and Uber's jaeger.

My main questions:

* Should the packages include the `Contrib` part? I don't have it right now because to me it seems more readable without it (e.g. `OpenTracing.Instrumentation.AspNetCore` vs. `OpenTracing.Contrib.Instrumentation.AspNetCore`)
* Should the tracer-packages be called `OpenTracing.Tracer.*`? There's the go-tracer "[Tracer](https://github.com/tracer/tracer)" so this might be problematic?! (especially if we have a `OpenTracing.Tracer.Tracer` package one day). PS: I don't think "tracer" is a good product name for a "tracer" :)
* Should tracer libraries be in the opentracing-contrib repo or should they be in the tracer repos? @adriancole Are you interested in hosting this package in a github.com/openzipkin repo with me as an owner?
* Should `OpenTracing.Tracer.Abstractions` be moved into the `opentracing` org? Java's AbstractTracer is there as well. Also, it would make sense to use this package for the BasicTracer implementation which currently is in the `opentracing` org as well.

Thanks for your time and feedback!

**Comment by adriancole** [comment]
wrt the "zipkin tracer" thing (combination of inject/extract and reporting I suspect?).. initially it is probably easiest to keep it here as that would eliminate some version drift as the thing evolves.

There are some incompatibility issues (like the recent dict+timestamp logging api), which require common tactics that should be applied similarly across zipkin implementations. Basically, success occurs when folks watching the code have equal stake in OpenTracing and Zipkin, as for example, things like dict logging are a mess otherwise (people dumping junk in zipkin as they don't care since they use something else).

Long story short, the best example OpenTracing and Zipkin working is `zipkin-opentracing-go`, led by @basvanbeek. If we had something similar in openzipkin for csharp/.net, it would have the same chance of success, not just now, but as apis continue to drift. If you are interested in championing similarly and/or inviting others in zipkin and/or bas, then I'd be happy to make a repo in openzipkin for this, and give you whatever access you need.

**Comment by bensigelman** [comment]
> Should the packages include the Contrib part? I don't have it right now because to me it seems more readable without it (e.g. OpenTracing.Instrumentation.AspNetCore vs. OpenTracing.Contrib.Instrumentation.AspNetCore)

I have a slight pref for including `contrib` in the artifact names, but that's just me. It will make it clearer (in some future scenario) which (few) artifacts are "core".

> Should the tracer-packages be called OpenTracing.Tracer.*? There's the go-tracer "Tracer" so this might be problematic?! (especially if we have a OpenTracing.Tracer.Tracer package one day). PS: I don't think "tracer" is a good product name for a "tracer" :)

They're just there as helpers, right? Maybe (???) `OpenTracing.TracerUtil.*`? Just a thought. What you have now is also fine with me.

(And re your PS: I agree!)

> Should tracer libraries be in the opentracing-contrib repo or should they be in the tracer repos? @adriancole Are you interested in hosting this package in a github.com/openzipkin repo with me as an owner?

I think it's more sensible to have an OT-Zipkin Tracer live in a zipkin repo, but that's ultimately up to Zipkin (and Adrian has weighed in, of course).

> Should OpenTracing.Tracer.Abstractions be moved into the opentracing org? Java's AbstractTracer is there as well. Also, it would make sense to use this package for the BasicTracer implementation which currently is in the opentracing org as well.

I really dislike where AbstractTracer ended up in Java, so I'd rather have them in contrib, at least for now. Having them in the core git repo implies a level of stability I don't think is appropriate.

Thanks! (And sorry for the delay)

**Comment by TerribleDev** [comment]
@cwe1ss 👍

**Comment by cwe1ss** [comment]
I moved my instrumentation code for ASP.NET Core into the new [opentracing-contrib/csharp-aspnetcore project](https://github.com/opentracing-contrib/csharp-aspnetcore). I'd like to invite everyone who's interested to contribute there as well!

I'm especially interested in feedback on 
https://github.com/opentracing-contrib/csharp-aspnetcore/issues/2 which discusses the package/library names for nuget.org
